### PR TITLE
llvm: Fix source build with modern GCC on Linux.

### PR DESCRIPTION
**Pull request by chandlerc**
Modern GCCs on Linux issue a warning that breaks the build that is
disabled with a flag next to `-fpermissive`. This warning doesn't
trigger upstream, but there isn't a point release with this fix.

Also, modern GCCs / linkers will build parts of Polly without PIC and
then complain about linking it into the DSO. However, PIC has very low
cost on Linux and so a simple workaround is just to fully enable a PIC
build. Given that the build already uses both an LLVM and Clang DSO to
reduce the install size, this has minimal impact and mostly fixes the
linking of Polly. It's unclear why this link ever succeeded when
building a bottle, but may reflect system-wide defaults for using PIC.
At least on a Debian testing Linux machine, this is necessary for me to
get the build from source to succeed.

-----

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/llvm.rb [hunk 1]
```diff
@@ -17,7 +17,6 @@ class Llvm < Formula
     sha256 cellar: :any,                 big_sur:       "ff9a71b7b35ecb6c1dfcfe40152b00f4777a3f4a10dcf5cc41044458b02c99cd"
     sha256 cellar: :any,                 catalina:      "e9a8185649b019863424068c3efa1b3c7d85747d12260b1dac07895690f50383"
     sha256 cellar: :any,                 mojave:        "2382196cb9626c46aa27ce146270c6625deca1b8a47926457131fb22ba243899"
-    sha256 cellar: :any_skip_relocation, x86_64_linux:  "ef3cc9c115ec2feb8ad2c317ddbfe190ad273a39acd737cd695fd700fb175c97"
```
Comment (jonchang) [inline Formula/llvm.rb:20]: Please restore this line; test bot automation will update it for you.

Comment (chandlerc) [reply Formula/llvm.rb]: Done, and sorry. I did this based on the instructions in the old linuxbrew repository, I'll try to internalize the newer instructions.

**Comment by danielnachun** [comment]
Great to get another contribution from you @chandlerc!  We really improved our LLVM bottles with your help so far.

Should we consider adding another test that tries to link to `libPolly`?  I know our test is already very extensive and I don't want to make it unmanageable, but right now it seems we only test `clang` and the run time libraries and perhaps we need to also test things like `libPolly` and `libLLVM`.  This is just a thought for now but I'm happy to discuss.

P.S. this is not related to this PR, but I did discover one last undesired use of GCC runtime libs in some parts of `compiler-rt`.  I'll address this in another PR soon and tag you in it.

**Comment by carlocab** [comment]
FYI, this will probably get killed when our macOS runners are taken offline for an upgrade later today.

I'll probably also end up cherry-picking your changes into https://github.com/Homebrew/homebrew-core/pull/83017 to avoid doubling up on CI runs.

**Comment by chandlerc** [comment]
In a bug report, it was suggested to also bundle a ClangD fix: https://github.com/Homebrew/homebrew-core/issues/84365

Let me know how you all would like this bundled... I'll hopefully get more time to poke at this later this week.

**Comment by carlocab** [comment]
https://github.com/llvm/llvm-project/commit/ec1fb95 applies with an offset, so I think we can just add that patch here.

**Comment by chandlerc** [comment]
> [llvm/llvm-project@ec1fb95](https://github.com/llvm/llvm-project/commit/ec1fb95) applies with an offset, so I think we can just add that patch here.

Wow, that's a super neat way to import the patch. I love it. Pushed including this.

**Comment by carlocab** [comment]
CI isn't running for some reason... let's try to get it going.

**Comment by carlocab** [comment]
Rolled a bunch of the changes from #83017 into here. Apologies for riding on your commit, but @BrewTestBot would've squashed them all into one even if I added them separately anyway.

**Comment by cho-m** [comment]
We will need to rebuild `mesa` bottle on Linux after merging this. It looks like current `mesa` has linked `libelf` formula (`libelf.so.0`) in `libvulkan_radeon.so` due to LLVM.

**Review by carlocab — APPROVED** [review]

**Comment by carlocab** [comment]
We can probably just rev bump `mesa` to shake off the `libelf` linkage.

**Pull request closed without merging**